datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -841,22 +838,65 @@ public void setDatastreamTaskStateForKey(DatastreamTask datastreamTask, String k
   }
 
   /**
-   * Remove instance assignment nodes whose instances are dead
+   * Remove instance assignment nodes whose instances are dead. This also
+   * removes the node from under /connector if present in unusedTasks.
    *
    * NOTE: this should only be called after the valid tasks have been
    * reassigned or become safe to discard per strategy requirement.
    * Coordinator is expected to cache the "current" assignment before
    * invoking the assignment strategy and pass the saved assignment
    * to us to figure out the obsolete tasks.
+   * @param previousAssignmentByInstance Previous task assignment
+   * @param newAssignmentsByInstance New task assignment.
    * @param instances Instances for which nodes are to be cleaned up
    */
-  public void cleanupDeadInstanceAssignments(List<String> instances) {
-    List<String> deadInstances = getAllInstances();
+  public void cleanUpDeadInstanceDataAndOtherUnusedTasks(
+      Map<String, Set<DatastreamTask>> previousAssignmentByInstance,
+      Map<String, List<DatastreamTask>> newAssignmentsByInstance,
+      List<String> instances) {
+
+    LOG.info("cleanUpDeadInstanceData is called");
+    Set<DatastreamTask> unusedTasks = findOldUnusedTasks(previousAssignmentByInstance, newAssignmentsByInstance);
+
+    Set<String> deadInstances = previousAssignmentByInstance.keySet();
     deadInstances.removeAll(instances);
+
+    LOG.debug("unusedTasks before cleanUp: {}", unusedTasks.stream()
+        .collect(Collectors.toMap(DatastreamTask::getDatastreamTaskName, Function.identity())).keySet());
```

Comment:
If you're only interested in `keySet()` then maybe we don't need a map. A list will do.
```java
    LOG.debug("unusedTasks before cleanUp: {}", unusedTasks.stream()
        .map(DatastreamTask::getDatastreamTaskName).collect(Collectors.toList()));
```
The same applies to lines 910-911.